### PR TITLE
Primary Button Implementation

### PR DESCRIPTION
**Pull request by chemscn**
## Description
Created Primary Button Component

## Motivation and Context
This is the first component for the UI-library.

## How Has This Been Tested?
Ran locally.

## Screenshots (if appropriate):
image

## Types of changes

Patch
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation change (documentation only, will not require redeployment)
- [ ] Refactor or cleanup (functionality unchanged)

Minor
- [ ] New feature (non-breaking change which adds functionality)

Major
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation, and I have updated the documentation accordingly.
- [x] I have performed a self-review of my own code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My change requires a version increment in package.json
- [x] I have confirmed all acceptance criteria defined by the story have been met

### CODE REVIEW

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    cursor: not-allowed;
+    background-color: $color--disabled3;
+    box-shadow: none;
+    transition: none;
+  }
+}
+.btn {
+  font-size: 12px;
+  border-radius: 6px;
+  padding: $button--padding;
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss:28]: If we aren't going to be creating variables for the other properties like `border-radius` we can probably remove this variable.

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss [hunk 2]
```diff
@@ -0,0 +1,60 @@
+@import "../../styles/global.styles";
+@import "../../styles/variables/constants";
+
+$transition-out: ease-in 0.1s;
+$transition-in: ease-out 0.3s;
+$box-shadow-color: none;
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss:6]: This was a left over of mine. Can be removed.

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+@mixin boxShadow($color) {
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss:7]: Hey these mixins are pretty cool

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.html [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<button
+  [ngClass]="[
+    'btn',
+    modeType === mode.PRIMARY ? 'btn--primary' : '',
+    modeType === mode.SECONDARY ? 'btn--secondary' : '',
+    modeType === mode.TERTIARY ? 'btn--tertiary' : '',
+    modeType === mode.HIGHLIGHT ? 'btn--highlight' : ''
+  ]"
+  [attr.aria-disabled]="disabled ? true : null"
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.html:9]: **Suggestion:** Thinking we might want something more for disabled state. Add better context somehow. Add icon to button? We should create a ticket for this.

File: projects/instant-ui-library/src/lib/root/instant-ui-library.component.ts [hunk 5]
```diff
@@ -4,6 +4,6 @@ import { Component, ViewEncapsulation } from '@angular/core';
   selector: 'iu-root',
   template: `<ng-content></ng-content> `,
   styleUrls: ['../styles/_global.styles.scss'],
-  encapsulation: ViewEncapsulation.None,
+  encapsulation: ViewEncapsulation.ShadowDom,
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/root/instant-ui-library.component.ts:7]: 👍 didn't know this was in here.

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    box-shadow: 0px 8px 15px $color;
+  }
+}
+@mixin createButton($shadow-color, $text-color) {
+  background-color: $shadow-color;
+  color: $text-color;
+  @include boxShadow($shadow-color);
+  &:disabled {
+    cursor: not-allowed;
+    background-color: $color--disabled3;
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss:20]: Have a concern with contrast. Should create ticket for this.

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  @Input() text: string;
+  @Input() title: string;
+  @Input() type: 'submit' | 'reset' | 'button' = 'button';
+  @Input() modeType: ModeType = 'primary';
+  @Input() disabled: boolean = false;
+  @Output() onClick: EventEmitter<null> = new EventEmitter<null>();
+  mode = Mode;
+
+  ngOnInit(): void {
+    if (!this.title) {
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.ts:28]: Should add this piece to `ngOnChanges`

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  cursor: pointer;
+  border: none;
+  border-radius: 8px;
+  transition: $transition-out;
+  &:hover,
+  &:focus {
+    transition: $transition-in;
+  }
+  &:focus {
+    outline: 0.1px solid $color--black;
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss:40]: **Suggestion:** I think I can see outline related items to be reused. Thoughts on adding as a mixin or variable. Probably create a mixen that adds this whole `&:focus`

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+}
+.btn {
+  font-size: 12px;
+  border-radius: 6px;
+  padding: $button--padding;
+  font-weight: 700;
+  margin: 5px 5px;
+  cursor: pointer;
+  border: none;
+  border-radius: 8px;
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss:33]: **Suggestion:** Can see `border-radius` being reused in other elements such as inputs. Thoughts on variable? I can see benefit of having it controlled by the component directly as well like we have here.

**Review by jacarder — CHANGES_REQUESTED** [review]

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  templateUrl: './base-button.component.html',
+  styleUrls: ['./base-button.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class BaseButtonComponent implements OnInit {
+  constructor() {}
+  @Input() text: string;
+  @Input() title: string;
+  @Input() type: 'submit' | 'reset' | 'button' = 'button';
+  @Input() modeType: ModeType = 'primary';
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/base-button/base-button.component.ts:22]: Can use enum to set default?

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.html (hunk 4 above)
Comment (chemscn) [reply projects/instant-ui-library/src/lib/button/base-button/base-button.component.html:9]: [Discuss disabled state](https://trello.com/c/7CQa42ll/21-discuss-disabled-states)

File: projects/instant-ui-library/src/lib/shared/models/color.model.ts [hunk 11]
```diff
@@ -0,0 +1,7 @@
+export type ModeType = 'primary' | 'secondary' | 'tertiary' | 'highlight';
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/shared/models/color.model.ts:1]: `export type ModeType = Mode.PRIMARY | Mode.SECONDARY | Mode.TERTIARY | Mode.HIGHLIGHT;`

File: projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss (hunk 6 above)
Comment (chemscn) [reply projects/instant-ui-library/src/lib/button/base-button/base-button.component.scss:20]: [Update disabled state for button component](https://trello.com/c/QmcL0NaF/22-update-disabled-state-for-button-component)

File: projects/instant-ui-library/src/lib/button/models/button.model.ts [hunk 12]
```diff
@@ -0,0 +1 @@
+export type ButtonType = 'primary' | 'secondary' | 'tertiary';
```
Comment (jacarder) [inline projects/instant-ui-library/src/lib/button/models/button.model.ts:1]: Is the overlap between this and `color.model` wanted?

Comment (chemscn) [reply projects/instant-ui-library/src/lib/button/models/button.model.ts:1]: This is not needed anymore.

File: projects/instant-ui-library/src/lib/shared/models/color.model.ts (hunk 11 above)
Comment (jacarder) [reply projects/instant-ui-library/src/lib/shared/models/color.model.ts:1]: This was marked resolved?

Comment (chemscn) [reply projects/instant-ui-library/src/lib/shared/models/color.model.ts:1]: Oops

**Review by jacarder — APPROVED** [review]

**Pull request merged**